adoptedStyleSheets.js

```diff
@@ -5,6 +5,51 @@
     return;
   }
 
+  // Can we rely on document.body
+  let polyfillLoaded = false;
+
+  // Polyfill-level reference to the iframe body
+  let frameBody;
+
+   // Style elements that will be attached to the head
+   // that need to be moved to the iframe
+  const deferredStyleSheets = [];
+
+  // Initialize the polyfill — Will be called on the window's load event
+  function initPolyfill() {
+    // Iframe is necessary because to extract the native CSSStyleSheet object
+    // the style element should be connected to the DOM.
+    const iframe = document.createElement('iframe');
+    iframe.hidden = true;
+    document.body.appendChild(iframe);
+
+    frameBody = iframe.contentWindow.document.body;
+
+    // Since we get the sheet from iframe, we need to patch prototype of the
+    // CSSStyleSheet in iframe as well.
+    updatePrototype(iframe.contentWindow.CSSStyleSheet.prototype);
+
+    // Document body will be observed from the very start to catch all added
+    // custom elements
+    createObserver(document.body);
+
+    // Document has loaded
+    polyfillLoaded = true;
+
+    // Move style elements created before document.body
+    // to the iframe along with future styles
+    deferredStyleSheets.forEach(nativeStyleSheet => {
+      frameBody.append(nativeStyleSheet);
+      nativeStyleSheet.disabled = false;
+    });
+
+    // Clear out the deferredStyleSheets array
+    deferredStyleSheets.length = 0;
+  }
+
+  // Proceed with using the iframe to house style elements
+  window.addEventListener('DOMContentLoaded', initPolyfill);
```

Comment:
Ok, that wasn't Safari, that was me 🙄

First of all, there should be
```suggestion
  document.addEventListener('DOMContentLoaded', initPolyfill);
```
Also, it works only if the `<script>` is in `<head>`. Otherwise, the event is fired long before we subscribe to it, and polyfill isn't initialized. So we need to
```suggestion
  if (document.body) {
    initPolyfill();
  } else { 
    document.addEventListener('DOMContentLoaded', initPolyfill);
  }
```
And move it with the `initPolyfill` function to the bottom of the polyfill, after the line:
```javascript
window.CSSStyleSheet = ConstructStyleSheet;
```
It would allow to avoid `cannot use element before initialization` errors when the `initPolyfill` uses the functions defined after it.

The `load` event of `window` will work as well, but I would suggest not to use it, because it may cause mistakes in the order of execution, when some scripts that use the polyfill are executed while it waits for all scripts loaded.

Reply:
With these changes, all tests in Safari and FF are green for me (Safari 12.1, Mojave)

Reply:
Hmm, steps to reproduce red cases? I just launched tests via `npm run test:watch` and got a lot of test failures. I assume it happens because Karma loads scripts to the body when the `DOMContentLoaded` script is already fired.